### DISCUSSION TITLE
How Can I Prevent Snapshotting Corrupted Data in Automated Backup Systems?

### DISCUSSION THREAD

**Discussion by kant777** [longhorn/longhorn#9645] [General]
Suppose I set up a periodic job to take volume snapshots every hour. If the data becomes corrupted (due to power loss or other factors) and a snapshot is taken shortly after, the snapshot will capture the corrupted data without me realizing it, since it’s an automated job. Now, if I try to restore from that snapshot a month later, I’d have to revert to a much earlier point in time, before the corruption occurred. What are the standard practices to handle this situation?

**What is corrupted data?**

By "corrupted data," I mean when an application attempts to write data to disk using a filesystem, and the filesystem writes data in blocks. If there's a power loss while a block is being written, it could result in incomplete or zero data being written to disk

**Potential solution:**

one potential solution is if the disk is being mirrored using RAID then filesystems like ZFS can perform integrity checks block by block across the mirrored disks and automatically fix the block with corrupted data

**Potential feature request**

If we need to constantly do integrity checks and perform recovery's is there a way to automate this process using kuberentes as a standard practice? If so, what are steps that needs to be taken?

**Comment by c3y1huang** [comment]
>  If the data becomes corrupted (due to power loss or other factors) and a snapshot is taken shortly after, the snapshot will capture the corrupted data without me realizing it

> By "corrupted data," I mean when an application attempts to write data to disk using a filesystem, and the filesystem writes data in blocks. If there's a power loss while a block is being written, it could result in incomplete or zero data being written to disk

Could you please clarify this in a bit more detail? Are you referring to corruption at the application (pod) filesystem layer?

**Reply by kant777** [reply]
In general I mean all possible ways where the data can be corrupted while writing to a disk (be it from application pod FS or the host FS) at some point I would imagine it needs to be written to a disk.

**Reply by voarsh2** [reply]
> In general I mean all possible ways where the data can be corrupted while writing to a disk (be it from application pod FS or the host FS) at some point I would imagine it needs to be written to a disk.

If you want guarantees, I'd use something like Ceph. She may be slower, but I haven't suffered like I have trying to use Longhorn in production.

**Reply by innobead** [reply]
@voarsh2 What kind of issues you encountered before when using Longhorn? What version did you use?
